doc/lab.1
.\" Automatically generated by Pandoc 3.1.11
.\"
.TH "LAB" "1" "January 14, 2024" "lab 1.0.0" "User Manual"
.SH NAME
lab \- A program for conducting CIT labs
.SH SYNOPSIS
\f[B]lab\f[R] SUB\-COMMAND [\f[I]args\f[R]]
.PP
SUB\-COMMAND := {start | check | status | current | flag | submit |
score | help}
.SH DESCRIPTION
\f[B]lab\f[R] is a CIT lab system, built to guide students through lab
assignments, providing feedback and tips, tracking progress and giving
students additional experience using the Unix command line.
.PP
Each lab consists of multiple \f[B]steps\f[R], and each step composed of
one or more \f[B]tasks\f[R].
Tasks will be either a command that must be typed exactly as shown, or a
command that is described by that students must figure out.
After completing all of the tasks in the step, students can check to see
if they have performed them properly, and to try again if not.
Once a step has been checked successfully, students may move on to the
next step.
After all steps are completed, the student will submit the lab for
grading.
This will provide a \f[B]submission code\f[R] that the student will
submit to Brightpsace.
.PP
Labs may also contain \f[B]flags\f[R].
A \f[B]flag\f[R] is a four digit number that appears after the text
\[lq]flag_\[rq].
For example: \[lq]flag_1234\[rq].
These will appear when running certain commands.
Students may \[lq]capture the flag\[rq] by using a command.
Some flags will appear as long as the student executes the steps
properly.
Other flags, called \f[B]bonus flags\f[R], are hidden and require the
student to explore and run commands that are not part of the lab.
These are worth bonus points.
.PP
\f[B]lab\f[R] also provides sub\-commands for viewing the steps of the
lab, the details of the current step
.SH SCORING
The score is calculated by adding the number of successfully completed
steps to the number of captured flags and the number of bonus flags.
The highest possible score is the number of steps plus the number of
flags.
This means that if you find bonus flags, your score could potentially be
higher than the total possible.
.PP
For example, if a lab has 7 steps, has 3 flags and 2 bonus flags, the
total possible is:
.IP
.EX
7 + 3 => 10 total possible points
.EE
.PP
If you capture the 2 bonus flags, your score can be 12 out of 10.
Those two points will be bonus points toward your final grade in the
class.
.SH SUB\-COMMANDS
.TP
\f[B]help\f[R]
display help message
.TP
\f[B]start\f[R] <\f[I]lab\-id\f[R]>
starts a new lab.
When you run the start command, you must provide the lab id (see
EXAMPLES), such as \[lq]2\-1\[rq], which will be given to you in the
assignment description in Brightspace.
Only one lab may be in progress at a time.
If you try to start a lab before submitting one that is in progress, you
will get a notice and the opportunity to submit the previous lab.
.TP
\f[B]steps\f[R]
view a list of steps for the currently active lab.
.TP
\f[B]current\f[R]
view the details and tasks of the current step.
.TP
\f[B]check\f[R]
check to see if you have completed the current step properly.
If not, you may try again and continue checking until you succeed.
Upon success, this command will prompt you to move to the next step if
you are ready.
.TP
\f[B]flag\f[R] <\f[I]flag\-number\f[R]>
captures a flag or bonus flag.
This command requires a four digit flag number argument (see EXAMPLES)
that will be seen in the output of a command during the lab (regular
flags) or while exploring on your own (bonus flags).
This can only capture flags that are valid for the active lab.
.TP
\f[B]submit\f[R]
submit the lab for grading.
You can submit a lab without completing all of the steps or capturing
all regular flags.
Your score report will be printed to the screen, along with a unique
\f[B]submission code\f[R] that you will submit to Brightspace to verify
that you have completed the lab.
.TP
\f[B]score\f[R]
review the score and results of a submitted lab.
This will display the number of completed steps, captured flags and
bonus flags and total score.
.SH EXAMPLES
.TP
\f[B]lab start 5\-2\f[R]
Begins lab number 5\-2.
.TP
\f[B]lab flag 1234\f[R]
Captures flag number 1234, if it is a valid flag for the current lab.
.SH AUTHORS
Written by Jeremy Doolin \c
.MT jdoolin@wvncc.edu
.ME \c
\&.
.SH BUGS
Submit bug reports to Jeremy Doolin \c
.MT jdoolin@wvncc.edu
.ME \c


lab.go
package ulab

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
)

type StepTest struct {
	TestType  string `json:"testType"`
	Command   string `json:"command"`
	Condition bool   `json:"condition"` // do we test for true/pass or false/fail?
}

type Step struct {
	Text           string   `json:"text"`
	Tasks          []string `json:"tasks"`
	Tips           string   `json:"tips"`
	Test           StepTest `json:"test"`
	SuccessMessage string   `json:"successMessage"`
	RetryMessage   string   `json:"retryMessage"`
}

type Lab struct {
	Number        string `json:"number"`
	Name          string `json:"name"`
	Description   string `json:"description"`
	Datafiles     bool   `json:"dataFiles"`
	Steps         []Step `json:"steps"`
	SubmitMessage string `json:"submitMessage"`
	Flags         []int  `json:"flags"`
	BonusFlags    []int  `json:"bonusFlags"`
}

func (l *Lab) Extract() {
	if !l.Datafiles {
		fmt.Printf("No data files are required for this lab.")
	} else {
		fmt.Printf("\nThis lab requires data files.\n")
		fmt.Printf("Extracting data files...")
		datafilePath := ULConfig.Root + "/labs/" + l.Number + "/data.zip"
		_, err := exec.Command("/usr/bin/unzip", datafilePath).Output()
		if err != nil {
			fmt.Printf("Error extracting data files: %v\n", err)
		}
	}
}

func (l *Lab) PrintSteps(s *Status) {
	results := s.GetResults(l.Number)
	fmt.Printf("Lab %s Steps\n\n", l.Number)
	fmt.Printf("   Flags: %d	Bonus Flags: %d\n", len(l.Flags), len(l.BonusFlags))
	fmt.Printf("   Steps:\n")
	for i, step := range l.Steps {
		stepStatus := "incomplete"
		if i < len(results.Steps) {
			stepStatus = results.Steps[i]
		}
		fmt.Printf("      %d. %s (%s)\n", i+1, step.Text, stepStatus)
	}
}

func (l *Lab) Check(step int) bool {
	test := l.Steps[step].Test
	switch test.TestType {
	case "script":
		// execute check command
		cmdPath := ULConfig.Root + "/labs/" + l.Number + "/scripts/" + test.Command
		//fmt.Printf("Running script %s...\n", cmdPath)
		_, err := exec.Command("sh", cmdPath).Output()
		// fun new switch statement
		switch {
		case (err != nil && test.Condition == true), (err == nil && test.Condition == false):
			fmt.Printf("%v\n", err)
			fmt.Printf("\n%s\n\n", l.Steps[step].RetryMessage)
			return false
		case (err == nil && test.Condition == true), (err != nil && test.Condition == false):
			fmt.Printf("\n%s\n\n", l.Steps[step].SuccessMessage)
			return true
		}
	case "command":
		// execute check command
		//fmt.Printf("Running command %s...\n", test.Command)
		_, err := exec.Command("sh", "-c", test.Command).Output()
		switch {
		case (err != nil && test.Condition == true), (err == nil && test.Condition == false):
			fmt.Printf("%v\n", err)
			fmt.Printf("\n%s\n\n", l.Steps[step].RetryMessage)
			return false
		case (err == nil && test.Condition == true), (err != nil && test.Condition == false):
			fmt.Printf("\n%s\n\n", l.Steps[step].SuccessMessage)
			return true
		}
		/*
			if err != nil {
				fmt.Printf("%v\n", err)
				fmt.Printf("\n%s\n\n", l.Steps[step].RetryMessage)
				return false
			} else {
				fmt.Printf("\n%s\n\n", l.Steps[step].SuccessMessage)
				return true
			}
		*/
	default:
		fmt.Printf("Not a recognized test type.")
		os.Exit(1)
		return false
	}
	return false
}

func (l *Lab) PrintStep(stepNum int) {
	l.Steps[stepNum].PrintTasks(stepNum)
}

func (s *Step) PrintTasks(stepNum int) {
	fmt.Printf("Step %d: %s\n\n", stepNum+1, s.Text)
	fmt.Println("Perform the following tasks/commands:")
	for _, task := range s.Tasks {
		fmt.Printf("\t%s\n", task)
	}
	fmt.Printf("\tlab check\n\n")
	if s.Tips != "" {
		fmt.Printf("TIPS: %s\n", s.Tips)
	}
}

func (l *Lab) CheckFlag(flagNum int) bool {
	for _, flag := range l.Flags {
		if flag == flagNum {
			return true
		}
	}
	return false
}

func (l *Lab) CheckBonusFlag(flagNum int) bool {
	for _, flag := range l.BonusFlags {
		if flag == flagNum {
			return true
		}
	}
	return false
}

func OpenLabFile(labNum string) *Lab {
	filepath := ULConfig.Root + "/labs/" + labNum + "/lab.json"
	//fmt.Printf("Opening lab file %s\n", filepath)
	labJson, err := os.ReadFile(filepath)
	if err != nil {
		fmt.Printf("Could not read data file for Lab %s: ", labNum)
		os.Exit(1)
		return nil
	} else {
		lab := Lab{}
		json.Unmarshal(labJson, &lab)
		return &lab
	}
}
